[trace] add Unmarshaler functionality to SpanContext and subfields

jackgopack4 opened the pull request:
Adds Unmarshaler interface implementations for `trace.SpanContext` and subfields.

This closes https://github.com/open-telemetry/opentelemetry-go/issues/6737 and also will enable less custom code to close https://github.com/open-telemetry/opentelemetry-collector/issues/11740 and https://github.com/open-telemetry/opentelemetry-collector/issues/12212.

Unit tests added for new functionality.

dmathieu commented:
This should close https://github.com/open-telemetry/opentelemetry-go/issues/1819

dmathieu commented on CHANGELOG.md:
```diff
@@ -15,6 +15,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
   The package contains semantic conventions from the `v1.31.0` version of the OpenTelemetry Semantic Conventions.
   See the [migration documentation](./semconv/v1.31.0/MIGRATION.md) for information on how to upgrade from `go.opentelemetry.io/otel/semconv/v1.30.0`(#6479)
 - Add `Recording`, `Scope`, and `Record` types in `go.opentelemetry.io/otel/log/logtest`. (#6507)
+- Add explicit JSON Unmarshaler interface methods for `trace.SpanContext` objects and subfields in `go.opentelemetry.io/otel/trace` (#6738).
```
```suggestion
- Add explicit JSON Unmarshaler interface methods for `SpanContext`, `TraceID`, `SpanID`, and `TraceFlags` in `go.opentelemetry.io/otel/trace`. (#6738)
```

dmathieu commented on trace/trace.go:
```diff
@@ -55,14 +68,30 @@ func (t TraceID) String() string {
+// It expects the JSON to be a string (as produced by MarshalJSON), parses
+// it via TraceIDFromHex, and replaces the receiver's contents.
+func (t *TraceID) UnmarshalJSON(data []byte) error {
+	// 1) Unmarshal the JSON payload into a Go string.
```
IMHO, those comments are a bit too much.

jackgopack4 replied:
I will pare them down to match more closely the other code around it

dmathieu commented on trace/trace.go:
```diff
@@ -55,14 +68,30 @@ func (t TraceID) String() string {
 
-var (
-	nilSpanID SpanID
-	_         json.Marshaler = nilSpanID
```
It would make more sense to keep the type validation alongside the definition than splitting things up.

jackgopack4 replied:
noted, thanks. have moved them.

dmathieu commented on trace/trace_test.go:
```diff
@@ -571,3 +572,253 @@ func TestConfigLinkMutability(t *testing.T) {
+		},
+		{
+			name:    "invalid JSON",
+			input:   `{"TraceID":123}`,
```
This is an invalid trace ID, but it's not invalid json.

```suggestion
			input:   `{"TraceID:123}`,
```

jackgopack4 replied:
good catch, thank you. fixed.

pellared commented:
As far as  I remember, we never had an agreement that we should add support for unmarshaling.

jackgopack4 commented:
> As far as I remember, we never had an agreement that we should add support for unmarshaling.

Thanks for this context; I work more in the collector than in the library so I was not aware of previous discussions regarding this matter. I just didn't understand why the data would support marshaling but not unmarshaling.

Additionally, it would help us with collector stability, specifically the [observability requirements](https://github.com/open-telemetry/opentelemetry-collector/blob/main/docs/component-stability.md#observability-requirements) as the exporters create an internal span for tracking progress of requests through the pipeline via the `trace.SpanContext` object, which is lost if the traces are sent to the persistent queue via file storage extension or similar. 

We should be able to work around this by implementing custom structs and helpers to imitate the unmarshal function, but it made sense to me to submit the PR for review in the library as well.

pellared commented:
@jackgopack4 this was not a complain towards your work. Some of the conversation can be found here https://github.com/open-telemetry/opentelemetry-go/issues/1819.

I remember we did not want to add capabilities to SDK that are not defined in the specification and could be a problem in future e.g. if an unmarshaling from JSON feature is defined in the specification. The fact that we have MarshalJSON in these exported types we consider as a legacy.

jackgopack4 commented:
thank you. If spec compliance is the concern, I've updated the JSON tags for the objects in this PR to match the OTel API Specification and added another note to the CHANGELOG to reflect it

jackgopack4 commented:
We were able to get what we needed done in collector using pdata and protobuf; so this dependency is no longer needed
